woocommerce/woocommerce-ios
# [Mobile Payments] Move Tap to Pay to payment method screen

Closes: #9234


## Description


This adds the Tap to Pay on iPhone payment acceptance method to the Payment Methods screen, and so splits the previous `Card` option in two: the other is now `Card Reader`.

Tapping on either card acceptance method will check for a connected reader, and disconnect the reader if it's of the opposite type. In that case, or when there's no reader connected, it will start a new connection to the correct type of reader, and then continue with the payment.

If the correct type of reader is already connected, it will immediately continue with the payment.

## Testing instructions


In a US-based store, using an iPhone XS or newer on iOS 16 and above:

Navigate to `Menu > Payments > Collect Payment`
Take a payment, selecting `Tap to Pay on iPhone` at the Payment Methods screen.
Observe that the built in reader connection flow starts, and is used to take the payment
Repeat this with a new payment
Observe that the connection flow doesn't start, it goes directly in to the payment using the built in reader (since you selected TTP)
Repeat this with a new payment, but select `Card Reader`
Observe that the Bluetooth card reader connection flow starts, and is used to take the payment


Repeat back and forth as much as required. 

It's also worth testing that the `Set up Tap to Pay on iPhone` and `Manage Card Reader` options work as expected: they also use this code.

Some interplay with the new caching of Onboarding caused issues towards the end of this PR. It is (still) possible to force the app in to a state where it doesn't realise its onboarding cache is out of date, and it can't take payments because of that issue. I don't believe that's related to this PR, it's part of the decision to move to caching that value, but still not checking onboarding when there's a reader connected, or while we're still in the Order pages.


## Screenshots


https://user-images.githubusercontent.com/2472348/227035400-65ef0370-bae9-4f08-aeab-a17ff32d3209.mp4



---
